File: PO/Project/JavaProject/src/Abstractions/World.java
package Abstractions;

import Animals.*;
import Plants.*;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.io.*;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

public class World extends JPanel implements ActionListener {
    private List<Organism> members;
    private Field[][] fields;
    private Dimension dimensions;

    private Dimension window_size;
    private Dimension fieldSize;
    private Dimension padding;
    private int round;
    private boolean end;
    private int updateOrder;
    private LogManager logger;
    private int windowWidth;
    private int windowHeight;
    private JFrame frame;
    private KeyHandler keyHandler;

    private char playerDirection;

    public World(String filepath) {
        this(12, 12);
        load(filepath);
    }

    public World(int y, int x) {
        this.dimensions = new Dimension(x, y);
        this.fieldSize = new Dimension(30, 50);
        this.padding = new Dimension(10, 10);
        this.window_size = new Dimension(dimensions.width * (fieldSize.width + padding.width) + 100, dimensions.height * (fieldSize.height + padding.height) + 100);
        this.members = new ArrayList<>();
        this.round = 0;
        this.end = false;
        this.updateOrder = 0;

        this.frame = new JFrame("World Simulation");
        this.frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        this.frame.setSize(new Dimension(window_size.width + 480, window_size.height));
        this.frame.setLocationRelativeTo(null);
        //this.gamePanel = new GamePanel(x, y);
        //this.frame.add(gamePanel);
        //gamePanel.startGameThread();
        //this.frame.pack();

        // Set layout for the frame
        this.frame.setLayout(new BorderLayout());

        this.logger = new LogManager( 50, window_size.height / 240);
        this.frame.add(logger, BorderLayout.EAST);
        this.frame.add(new JScrollPane(logger), BorderLayout.EAST);
        this.frame.add(this, BorderLayout.CENTER);

        this.keyHandler = new KeyHandler();
        frame.addKeyListener(keyHandler);

        this.frame.setVisible(true);
        this.frame.setFocusable(true);
        this.frame.revalidate();
        this.frame.repaint();

        initializeFields();
    }


    private void initializeFields() {
        logger.log("Initialising fields...");
        this.fields = new Field[dimensions.height][dimensions.width];
        for (int i = 0; i < dimensions.height; i++) {
            for (int j = 0; j < dimensions.width; j++) {
                Point position = new Point(
                        j * (fieldSize.width + padding.width),
                        i * (fieldSize.height + padding.height)
                );
                fields[i][j] = new Field(position, fieldSize, new Point(j, i), null);
            }
        }

        logger.log("Initialised!\nAdding the player...");
        add(new Player(this, new Point(0,0)));

        logger.log("Added!\nGenerating random start...");
        generateRandomStart((dimensions.width * dimensions.height) / 12);

        logger.log("Generated!\nSorting members...");
        sortMembers();
        logger.log("Sorted!");
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        draw(g);
    }

    private void draw(Graphics g) {
        drawInterface(g); //?
        for (Field[] row : fields) {
            for (Field field : row) {
                field.drawBox(g);
                if (field.member != null) {
                    field.member.draw(g, field.position, fieldSize);
                }
            }
        }
    }

    private void drawInterface(Graphics g) {
//        g.drawString("Round: " + round, 200, 20);
//        g.drawString("Yauheni Pyryeu 201253 project", 200, 40);
//        g.drawString("q - quit, e - ability, w a s d - controls", 10, 200);
        System.out.println("Round " + (round + 1));
    }

    public void actionPerformed(ActionEvent e) {
        // Handle game actions here
    }

    public void update() {
        round++;
        for (int i = 0; i < members.size(); i++) {
            if (members.size() <= 1) {
                end = true;
            }
            if (end) {
                drawEndscreen();
                return;
            }
            if (updateOrder > 0 && updateOrder < members.size()) {
                i = updateOrder;
                updateOrder = 0;
            }

//            if (members.get(i) instanceof Player) {
//                // Wait for player's action
////                JOptionPane.showMessageDialog(null, "Player's turn!");
////                clean();
////                repaint();
//
////                try {
////                    Thread.sleep(50);
////                } catch (InterruptedException e) {
////                    e.printStackTrace();
////                }
////                playerDirection = ' ';
//                Player player = (Player) members.get(i);
//                try {
//                    player.action(keyHandler);
//                } catch (InterruptedException e) {
//                    throw new RuntimeException(e);
//                }
//            }
//            else {
                members.get(i).action();
//            }
        }
        clean();
        repaint();
    }

    public void add(Organism added) {
        added.setBirth(round);
        members.add(added);
        findField(added.getPosition()).setMember(added);
        sortMembers();
        logger.logCreation(added);
    }

    public void destroy(Organism destroyed) {
        logger.necroLog(destroyed);
        findField(destroyed.getPosition()).setMember(null);
        members.remove(destroyed);
        sortMembers();
    }

    public void sortMembers() {
        members.sort(Comparator.comparingInt(Organism::getInitiative)
                .thenComparingInt(Organism::getBirth).reversed());
    }

    public void stop() {
        this.end = true;
    }

    public boolean isOver() {
        return end;
    }

    public Dimension getDimensions()
    {
        return dimensions;
    }

    public Field findField(Point id) {
        if (id.y < dimensions.height && id.y >= 0 && id.x < dimensions.width && id.x >= 0) {
            return fields[id.y][id.x];
        } else {
            return null;
        }
    }

    public List<Field> getFieldsNear(Point position){
        List<Field> near = new ArrayList<>();

        for (int i = -1; i < 2; i++) {
            for (int j = -1; j < 2; j++) {
                if ((i == 0) && (j == 0)) {
                    continue;
                }

                Point target_point = new Point(position.x + j, position.y + i);
                Field target = findField(target_point);

                if (target != null) {
                    near.add(target);
                }
            }
        }

        return near;
    }

    public int getRound() {
        return round;
    }

    public LogManager getLogger() {
        return logger;
    }
    public KeyHandler getKeyHandler(){
        return keyHandler;
    }

    public Field getRandomField() {
        List<Field> freeFields = new ArrayList<>();
        for (Field[] row : fields) {
            for (Field field : row) {
                if (field.getMember() == null) {
                    freeFields.add(field);
                }
            }
        }
        if (freeFields.isEmpty()) {
            return null;
        } else {
            return freeFields.get(ThreadLocalRandom.current().nextInt(freeFields.size()));
        }
    }

    public Field getFreeFieldNear(Point position) {
        List<Field> freeFields = new ArrayList<>();
        for (int i = -1; i < 2; i++) {
            for (int j = -1; j < 2; j++) {
                if (i == 0 && j == 0) continue;
                Field target = findField(new Point(position.x + j, position.y + i));
                if (target != null && target.member == null) {
                    freeFields.add(target);
                }
            }
        }
        if (freeFields.isEmpty()) {
            return null;
        } else {
            return freeFields.get(ThreadLocalRandom.current().nextInt(freeFields.size()));
        }
    }

    public boolean sameType(Organism a, Organism b) {
        // Check if both references are non-null
        if (a != null && b != null) {
            // Compare the runtime types of the objects
            return a.getClass() == b.getClass();
        }
        return false;
    }

    public void generateRandomStart(int numberOfOrganisms) {
        for (int i = 0; i < numberOfOrganisms; i++) {
            generateRandomOrganism(ThreadLocalRandom.current().nextInt(1, 3), ThreadLocalRandom.current().nextInt(1, 6));
        }
    }

    public void generateRandomOrganism(int type, int subType) {
        Field randomField = getRandomField();
        if (randomField == null) {
            System.out.println("There is no free field!");
            return;
        }
        Organism organism = switch (type) {
            case 1 -> switch (subType) {
                case 1 -> new Wolf(this, randomField.id);
                case 2 -> new Sheep(this, randomField.id);
                case 3 -> new Fox(this, randomField.id);
                case 4 -> new Turtle(this, randomField.id);
                case 5 -> new Antelope(this, randomField.id);
                default -> null;
            };
            case 2 -> switch (subType) {
                case 1 -> new Grass(this, randomField.id);
                case 2 -> new Mlecz(this, randomField.id);
                case 3 -> new Guarana(this, randomField.id);
                case 4 -> new Wolfberry(this, randomField.id);
                case 5 -> new Hogweed(this, randomField.id);
                default -> null;
            };
            default -> null;
        };
        if (organism != null) {
            add(organism);
        }
    }

    public void drawEndscreen() {
        JOptionPane.showMessageDialog(frame, "End of Game! Round: " + round + (getPlayer() != null ? " You have won!!" : " You have lost!"));
    }

    public Organism getPlayer() {
        for (Organism member : members) {
            if (member instanceof Player) {
                return (Player) member;
            }
        }
        return null;
    }

    public void clean() {
        for (int i = 0; i < members.size(); i++) {
            int neighbours = countNeighboursSameType(members.get(i));
            if (neighbours > 2) {
                logger.logOverpopulation(members.get(i), neighbours);
                destroy(members.get(i));
                i--;
            }
        }
    }

    public int countNeighboursSameType(Organism me) {
        int count = 0;
        for (int i = -1; i < 2; i++) {
            for (int j = -1; j < 2; j++) {
                if (i == 0 && j == 0) continue;
                Field target = findField(new Point(me.getPosition().x + j, me.getPosition().y + i));
                if (target != null && target.member != null && target.member.getClass() == me.getClass()) {
                    count++;
                }
            }
        }
        return count;
    }

    public void save(String path) {
        try {
            FileWriter writer = new FileWriter(path);
            writer.write("{\"round\":" + round + ",\"dimensions\":{\"height\":" + dimensions.height + ",\"width\":" + dimensions.width + "},\"members\":");
            String json = new Gson().toJson(members);
            writer.write(json + "}");
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void load(String path) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(path));
            JsonObject jsonObject = new Gson().fromJson(br, JsonObject.class);
            this.round = jsonObject.get("round").getAsInt();
            JsonObject dims = jsonObject.getAsJsonObject("dimensions");
            this.dimensions = new Dimension(dims.get("width").getAsInt(), dims.get("height").getAsInt());
            initializeFields();
            JsonArray array = jsonObject.getAsJsonArray("members");
            for (JsonElement element : array) {
                JsonObject member = element.getAsJsonObject();
                char avatar = (char)(member.get("avatar").getAsInt());

                Organism entity = null;
                Point new_position = new Point();
                new_position.x = member.get("x").getAsInt(); // Assuming 'x' represents the x-coordinate of the position
                new_position.y = member.get("y").getAsInt();

                switch (avatar)
                {
                    case 'P':
                        entity = new Player(this,new_position);
                        break;
                    case 'W':
                        entity = new Wolf(this,new_position);
                        break;
                    case 'F':
                        entity = new Fox(this,new_position);
                        break;
                    case 'T':
                        entity = new Turtle(this,new_position);
                        break;
                    case 'A':
                        entity = new Antelope(this,new_position);
                        break;
                    case 'S':
                        entity = new Sheep(this,new_position);
                        break;
                    case 'g':
                        entity = new Grass(this, new_position);
                        break;
                    case 'u':
                        entity = new Guarana(this, new_position);
                        break;
                    case 'm':
                        entity = new Mlecz(this, new_position);
                        break;
                    case 'w':
                        entity = new Wolfberry(this, new_position);
                        break;
                    case 'h':
                        entity = new Hogweed(this, new_position);
                        break;
                    default:
                        break;
                }

                members.add(entity);
                findField(entity.getPosition()).setMember(entity);
            }
            sortMembers();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public void clearWorld() {
        members.clear();
        frame.dispose();
    }

//    public static void main(String[] args) {
//        new World("path/to/your/file.json");
//    }
}

File: Sem2/PO/Project/C++/src/world.cpp
#include "world.h"
#include "animal.h"
#include "animals.h"
#include "logmanager.h"
#include "organizm.h"
#include "plant.h"
#include "plants.h"
#include "plants.h"
#include "player.h"
#include <cstddef>
#include <cstdlib>
#include <ctime> 
#include <thread>
#include <chrono>
#include <ncurses.h>
#include <vector>
#include <fstream>
//#include<unistd.h>


world::world(const char* filepath)
{
    dimensions.y = 12;
    dimensions.x = 12;

    initscr();
    start_color();
    use_default_colors();

    //initialasing color pairs for different types
    init_pair(1, COLOR_RED, -1);
    init_pair(2, COLOR_GREEN, -1);
    init_pair(3, COLOR_YELLOW, -1);

    int max_y, max_x;
    getmaxyx(stdscr, max_y, max_x);
    window_width = max_x * 0.6;
    window_height = max_y;

    //initialasing a world's window
    worldWindow = newwin(max_y, window_width, 0, 0);

    noecho(); 
    curs_set(0);

    //initialising random for the whole project
    srand(time(NULL));

    //Logger initialisation
    Logger = new logmanager("log.txt");

    round = 0;
    end = false;
    this->field_size = {3, 5};
    this->padding = {0, 1};

    YX starting_padding = {2, window_width/2 - (dimensions.x/2*(field_size.x + padding.x))};

    Load(filepath);

    fields = new field*[dimensions.y];

    for (int i = 0; i < dimensions.y; i++) {
        fields[i] = new field[dimensions.x];
    }

    // Initializing fields
    for (int i = 0; i < dimensions.y; i++) {
        for (int j = 0; j < dimensions.x; j++) 
        {
            fields[i][j].window = worldWindow;

            fields[i][j].field_size = field_size;

            fields[i][j].position.y = i * (field_size.y + padding.y) + starting_padding.y;
            fields[i][j].id.y = i;
            fields[i][j].position.x = j * (field_size.x + padding.x) + starting_padding.x;
            fields[i][j].id.x = j;

            fields[i][j].member = nullptr;
        }
    }
    
    update_order = 0;

    for (int i = 0; i < members.size(); i++) {
        fields[members[i]->GetPosition().y][members[i]->GetPosition().x].member = members[i];
        if (dynamic_cast<player*>(members[i]) != nullptr) {
            update_order = i;
        }
    }

    SortMembers();
}

world::world(int y, int x)
{
    initscr();
    start_color();
    use_default_colors();

    //initialasing color pairs for different types
    init_pair(1, COLOR_RED, -1);
    init_pair(2, COLOR_GREEN, -1);
    init_pair(3, COLOR_YELLOW, -1);

    int max_y, max_x;
    getmaxyx(stdscr, max_y, max_x);
    window_width = max_x * 0.6;
    window_height = max_y;

    //initialasing a world's window
    worldWindow = newwin(max_y, window_width, 0, 0);

    noecho(); 
    curs_set(0);

    //initialising random for the whole project
    srand(time(NULL));

    //Logger initialisation
    Logger = new logmanager("log.txt");

    round = 0;
    end = false;
    dimensions = {y, x};
    this->field_size = {3, 5};
    this->padding = {0, 1};

    YX starting_padding = {2, window_width/2 - (x/2*(field_size.x + padding.x))};

    fields = new field*[y];

    for (int i = 0; i < y; i++) {
        fields[i] = new field[x];
    }

    // Initializing fields
    for (int i = 0; i < y; i++) {
        for (int j = 0; j < x; j++) 
        {
            fields[i][j].window = worldWindow;

            fields[i][j].field_size = field_size;

            fields[i][j].position.y = i * (field_size.y + padding.y) + starting_padding.y;
            fields[i][j].id.y = i;
            fields[i][j].position.x = j * (field_size.x + padding.x) + starting_padding.x;
            fields[i][j].id.x = j;

            fields[i][j].member = nullptr;
        }
    }

    player* Player = new player(this, {0,0});
    Add(Player);

    //GenerateRandomStart((y * x) / 6);
    GenerateEvenStart((y * x) / 12);

    update_order = 0;

    SortMembers();
}


field* world::FindField(YX id)
{
    if ((id.y < dimensions.y) && (id.y >= 0) && (id.x < dimensions.x) && (id.x >= 0))
    {
        return &(fields[id.y][id.x]);
    }
    else
    {
        return nullptr;
    }
}

field* world::GetRandomField()
{
    std::vector<field*> free_fields;

    for (int y = 0; y < dimensions.y; y++)
    {
        for (int x = 0; x < dimensions.x; x++)
        {
            if (fields[y][x].member == nullptr)
            {
                free_fields.push_back(&(fields[y][x]));
            }   
        }
    }
    
    int n_free_fields = free_fields.size();

    if (n_free_fields == 0)
    {
        return nullptr;
    }
    else
    {
        return free_fields[ rand() % n_free_fields ];
    }
}

field* world::GetFreeFieldNear(YX position)
{
    std::vector<field*> free_fields;

    for (int i = -1; i < 2; i++) {
        for (int j = -1; j < 2; j++) {
            if ((i == 0) && (j == 0)) {
                continue;
            }

            field* target = FindField({position.y + i, position.x + j});

            if ((target != nullptr) && (target->member == nullptr)) 
            {
                free_fields.push_back(target);
            }
        }
    }

    if (free_fields.size() == 0) {
        return nullptr;
    }

    else {
        return free_fields[rand() % free_fields.size()];
    }
}

std::vector<field*> world::GetFieldsNear(YX position)
{
    std::vector<field*> near;

    for (int i = -1; i < 2; i++) {
        for (int j = -1; j < 2; j++) {
            if ((i == 0) && (j == 0)) {
                continue;
            }

            field* target = FindField({position.y + i, position.x + j});

            if (target != nullptr) 
            {
                near.push_back(target);
            }
        }
    }

    return near;
}

void world::Draw()
{
    wclear(worldWindow);

    DrawInterface();

    for (int i = 0; i < dimensions.y; i++) {
        for (int j = 0; j < dimensions.x; j++) 
        {
            fields[i][j].draw_box(field_size);
            if (fields[i][j].member != nullptr) 
            {
                fields[i][j].member->Draw({fields[i][j].position.y + fields[i][j].field_size.y/2, fields[i][j].position.x + fields[i][j].field_size.x/2});
            }
        }
    }
    return;
}

void world::DrawInterface()
{
    mvwprintw(worldWindow,1, window_width/2 - 5, "Round is %d", round);
    mvwprintw(worldWindow,0, window_width/2 - 15, "Yauheni Pyryeu 201253 project");
    mvwprintw(worldWindow,window_height - 1, window_height/2 - 14, "q - quit, e - ability, w a s d - controls");
    
    return;
}

void world::DrawEndscreen()
{
    wclear(worldWindow);
    //wclear(Logger->GetWindow());

    mvwprintw(worldWindow,0, window_width/2 - 6, "End of Game!");
    mvwprintw(worldWindow,1, window_width/2 - 5, "Round is %d", round);
    
    if ((members.size() == 1) && (GetPlayer() != nullptr)) {
        mvwprintw(worldWindow,2, window_width/2 - 7, "You have won!!");
    }
    else if (GetPlayer() == nullptr) {
        mvwprintw(worldWindow,2, window_width/2 - 7, "You have lost!");
    }
    else {
        mvwprintw(worldWindow, 3, window_width/2 - 3, "saved");
        Save("save.txt");
    }

    //Draw();
    Logger->LogF(TECHNICAL, "---!log ended!---");
    wrefresh(Logger->GetWindow());

    wgetch(worldWindow);

    return;
}

organizm* world::GetPlayer()
{
    for (int i = 0; i < members.size(); i++) {
        if (player* player_ptr = dynamic_cast<player*>(members[i])) {
            return player_ptr;
        }
    }

    return nullptr;
}

void world::Update()
{
    round++;

    for (int i = 0; i < members.size(); i++)
    {
        if (members.size() <= 1)
        {
            end = true;
        }
        
        if (end)
        {
            return;
        }

        if ((update_order > 0) && (update_order < members.size())) {
            i = update_order;
            update_order = 0;
        }

        members[i]->Action();
    }

    Clean();
}

// void world::SleepMiliseconds(int ms)
// {
//     clock_t goal = ms + clock(); // The end time

//     while (goal > clock());
//     return;
// }

void world::Add(organizm* added)
{
    added->SetBirth(round);
    members.push_back(added);
    FindField(added->GetPosition())->member = added;

    SortMembers();

    Logger->LogCreation(added);

    return;
}

void world::Destroy(organizm* destroyed)
{
    Logger->NecroLog(destroyed);

    FindField(destroyed->GetPosition())->member = nullptr;

    //auto it = std::find(members.begin(), members.end(), destroyed);
    organizm* it = nullptr;
    int index;

    for (int i = 0; i < members.size(); i++) {
        if (members[i] == destroyed) {
            it = members[i];
            index = i;
        }
    }

    if (it == nullptr) { return; }

    if (dynamic_cast<player*>(it) != nullptr)
    {
        end = true;
    }

    delete destroyed;
    members.erase(members.begin() + index);

    SortMembers();
}
void world::SortMembers()
{
    std::sort(members.begin(), members.end(), [](organizm* a, organizm* b)
    {
        int aInitiative = a->GetInitiative();
        int bInitiative = b->GetInitiative();
        int aBirthRound = a->GetBirth();
        int bBirthRound = b->GetBirth();
        
        if (aInitiative == bInitiative)
        {
            return aBirthRound > bBirthRound;
        }
        return aInitiative > bInitiative; 
    });

    //Logger->LogOrder(members);
}

int world::CountNeighboursSameType(organizm* me)
{
    int count = 0;

    for (int i = -1; i < 2; i++) {
        for (int j = -1; j < 2; j++) {
            if (((i == 0) && (j == 0)) || (FindField({me->GetPosition().y + i, me->GetPosition().x + j}) == nullptr)) {
                continue;
            }

            organizm* target = FindField({me->GetPosition().y + i, me->GetPosition().x + j})->member;

            if ((target != nullptr) && (SameType(target, me))) {
                count++;
            } 
        }
    }

    return count;
}

bool world::SameType(organizm* a, organizm* b)
{
    // Check if both pointers are valid
    if (a && b) {
        // Get the type_info objects
        const std::type_info& type1 = typeid(*a);
        const std::type_info& type2 = typeid(*b);

        // Compare the type_info objects
        return type1 == type2;
    }
    return false; 
}

void world::Clean()
{
    //Logger->Log("cleaning routine...");

    for (int i = 0; i < members.size(); i++) {
        int neighbours = CountNeighboursSameType(members[i]);
        if (neighbours > 2) {
            Logger->LogOverpopulation(members[i], neighbours);
            Destroy(members[i]);
            i--;
        }
    }

    return;
}

void world::GenerateRandomOrganizm(int type, int sub_type)
{
    field* random = GetRandomField();
    if (random == nullptr)
    {
        clear();
        wprintw(worldWindow,"There is no free field!");
        wgetch(worldWindow);
        return;
    }
    

    // int type = rand() % 2 + 1; // Generates 1 or 2
    // int sub_type = rand() % 5 + 1; // Generates from 1 to 5

    if (type == 1) // animal
    {
        switch (sub_type)
        {
        case 1:
            // wolf
            Add(new wolf(this, random->id));
            break;
        case 2:
            // sheep
            Add(new sheep(this, random->id));
            break;
        case 3:
            // fox
            Add(new fox(this, random->id));
            break;
        case 4:
            // turtle
            Add(new turtle(this, random->id));
            break;
        case 5:
            // antelope
            Add(new antelope(this, random->id));
            break;
        default:
            break;
        }
    }
    else if (type == 2) // plant
    {
        switch (sub_type)
        {
        case 1:
            // grass
            Add(new grass(random->id, this));
            break;
        case 2:
            // mlecz
            Add(new mlecz(random->id, this));
            break;
        case 3:
            // guarana
            Add(new guarana(random->id, this));
            break;
        case 4:
            // wolfberry
            Add(new wolfberry(random->id, this));
            break;
        case 5:
            // hogweed
            Add(new hogweed(random->id, this));
            break;
        default:
            break;
        }
    }
}


void world::GenerateRandomStart(int number_of_organizms)
{
    for (int object = 0; object < number_of_organizms; object++)
    {
        GenerateRandomOrganizm(rand() % 2 + 1, rand() % 5 + 1);   
    }

    return;
}

void world::GenerateEvenStart(int number_of_organizms)
{
    int type = 1;
    int subtype = 1;
    int count = 0;

    while (count < number_of_organizms)
    {
        type = (type == 1) ? 2 : 1;

        if (count % 2 == 0) 
        {
            subtype = (subtype >= 5) ? 1 : (subtype + 1);
        }

        GenerateRandomOrganizm(type, subtype);
        count++;
    }

    return;
}

void world::Save(const char* filepath)
{
    // Open the file for writing
    std::ofstream file(filepath, std::ios::trunc);
    if (!file.is_open()) {
        // Handle error if file cannot be opened
        Logger->LogF(DEATH, "---!error saving file!---\n");
        std::cerr << "Error: Unable to open save file for writing.\n";
        return;
    }

    // World "info"
    file << dimensions.y << " " << dimensions.x << "\n";
    file << round << "\n";

    //members
    file << members.size() << "\n\n";
    for (int i = 0; i < members.size(); i++) {
        file << members[i]->GetAvatar() << "\n";
        file << members[i]->GetPosition().y << ' ' << members[i]->GetPosition().x << '\n';
        file << members[i]->GetStrength() << '\n';
        file << members[i]->GetBirth() << "\n\n";
    }

    player* player_ptr = dynamic_cast<player*>(GetPlayer());
    file << player_ptr->GetCooldown();

    // Close the file
    file.close();

    Logger->LogF(TECHNICAL, "---!The game is saved!---\n");

    return;
}

void world::Load(const char* filepath)
{
    // Open the save file for reading
    std::ifstream file(filepath);
    if (!file.is_open()) {
        // Handle error if file cannot be opened
        Logger->LogF(DEATH, "---!error loading from file!---\n");
        std::cerr << "Error: Unable to open save file for reading.\n";
        return;
    }

    file >> dimensions.y >> dimensions.x >> round;
    round--;

    int n_entities;
    file >> n_entities;

    for (int i = 0; i < n_entities; i++) {
        char avatar;
        file >> avatar;

        YX e_position;
        int strength, birth;

        file >> e_position.y >> e_position.x;
        file >> strength;
        file >> birth;

        organizm* entity = nullptr;

        switch (avatar) 
        {
            case 'P':
                entity = new player(this,{e_position.y, e_position.x});
                break;
            case 'W':
                entity = new wolf(this,{e_position.y, e_position.x});
                break;
            case 'F':
                entity = new fox(this,{e_position.y, e_position.x});
                break;
            case 'T':
                entity = new turtle(this,{e_position.y, e_position.x});
                break;
            case 'A':
                entity = new antelope(this,{e_position.y, e_position.x});
                break;
            case 'S':
                entity = new sheep(this,{e_position.y, e_position.x});
                break;
            case 'g':
                entity = new grass({e_position.y, e_position.x},this);
                break;
            case 'u':
                entity = new guarana({e_position.y, e_position.x},this);
                break;
            case 'm':
                entity = new mlecz({e_position.y, e_position.x},this);
                break;
            case 'w':
                entity = new wolfberry({e_position.y, e_position.x},this);
                break;
            case 'h':
                entity = new hogweed({e_position.y, e_position.x},this);
                break;
            default:
                break;
        }

        if (entity != nullptr) {
            entity->SetStrength(strength);
            entity->SetBirth(birth);
            members.push_back(entity);
        }
    }

    int cooldown;
    file >> cooldown;

    player* player_ptr = dynamic_cast<player*>(GetPlayer());
    player_ptr->SetCooldown(cooldown - 1);

    // Close the file
    file.close();

    Logger->LogF(TECHNICAL, "---!The game is loaded!---\n");
}

world::~world()
{
    for (organizm* org : members) {
        delete org;
    }

    members.clear();

    for (int i = 0; i < dimensions.y; i++) 
    {
        delete[] fields[i];
    }

    delwin(worldWindow);
    
    delete[] fields;
    delete Logger;  
    endwin();
}
